Media Replace Flow: Add custom toggle support and fix button height

yogeshbhutkar opened the pull request:
Fixes: #68029 

## What?
This PR contains a patch for fixing spacing inconsistency in the background image selector button.

## How?
As the issue description mentions, the default size of `Toolbar Button` was updated to `compact`, whose styles are applied using the `is-compact` classname.

The `is-compact` classname contains a height of `32px` as mentioned here:
https://github.com/WordPress/gutenberg/blob/581de9f893ff63c156b118f7244211f417fe7543/packages/components/src/button/style.scss#L298

https://github.com/WordPress/gutenberg/blob/581de9f893ff63c156b118f7244211f417fe7543/packages/base-styles/_variables.scss#L99

Whereas, the heights described in `.components-dropdown` applied to its parent, is actually `36px` leading to a mismatch in the parent container and child's heights.

https://github.com/WordPress/gutenberg/blob/581de9f893ff63c156b118f7244211f417fe7543/packages/block-editor/src/components/background-image-control/style.scss#L26

https://github.com/WordPress/gutenberg/blob/581de9f893ff63c156b118f7244211f417fe7543/packages/block-editor/src/components/background-image-control/style.scss#L47

To fix the issue, the heights are updated inside `.components-dropdown` to now use `fit-content` instead of using the previously hardcoded value `36px`.

As an alternative, we can also hardcode the height to `32px`.

## File Changed:
`> packages/block-editor/src/components/background-image-control/style.scss`

## Testing Instructions
1. Go to the site editor.
2. Click on "Styles"
3. Click on "Background"
4. Confirm that the spacing for the button is now correct.

## Screenshots

|Before|After|
|-|-|
|![Screenshot 2024-12-17 at 10 38 17 AM](https://github.com/user-attachments/assets/48e3b41a-6463-4210-849a-2990a4c79c4e)|![Screenshot 2024-12-17 at 10 37 24 AM](https://github.com/user-attachments/assets/d6bcc79c-3409-4f08-8b1b-af9ee51ae03b)|

yogeshbhutkar committed:
Update background image control dropdown height to fit-content

yogeshbhutkar committed:
Replace ToolbarButton with Button in MediaReplaceFlow component

yogeshbhutkar committed:
Refactor MediaReplaceFlow to support dynamic button variants and update BackgroundImageControls to specify button variant

yogeshbhutkar committed:
Refactor MediaReplaceFlow to set default button variant to 'toolbar' and remove unused default variant

tyxla reviewed (commented):
My feedback from #68038 has been addressed, so I'll let @mirka review individually when she gets a chance.

mirka commented on packages/block-editor/src/components/background-image-control/style.scss:
```diff
@@ -23,7 +23,7 @@
 	.components-dropdown {
 		display: block;
-		height: 36px;
+		height: fit-content;
```
Can these height rules be removed entirely?

```suggestion
```

yogeshbhutkar replied:
Thanks for the review @mirka.

Yes, we can indeed remove the height rules. I've removed them in the latest commit. Here's a screenshot of the component after recent changes:

![Screenshot 2024-12-20 at 8 56 42 AM](https://github.com/user-attachments/assets/0805ed47-d6c3-4eb1-8c3d-00866fcaf05b)

It's working as expected on my end.

> This is good, and I think we can future-proof it even further by exposing the entire renderToggle as an optional prop, rather than try to abstract it with buttonVariant.

Great Point! I'll work on incorporating this, will update here once done.

mirka reviewed (commented):
This is good, and I think we can future-proof it even further by exposing the entire `renderToggle` as an optional prop, rather than try to abstract it with `buttonVariant`. My concern with an abstract approach like `buttonVariant` is that we'll need to alter the API every time we need something slightly different, like an icon button or another Button variant.

So for example in BackgroundImageControls I want to be able to do something like:

```jsx
<MediaReplaceFlow
  renderToggle={ ( props ) => (
    <Button
      { ...props }
      __next40pxDefaultSize
    />
  ) }
```

What do you think? (cc @WordPress/gutenberg-components)

Also as a separate issue, I noticed that the focus manipulation that `editMediaButtonRef` is being used to do is not relevant/appropriate anymore in these recent implementations. I'm looking at the [original commit](https://github.com/WordPress/gutenberg/commit/f58d0e3024209c63b52912e82454dd6785af8097) where that focus manipulation was added, and it seems like it was for focus restoration, which doesn't apply anymore (cc @draganescu in case I'm wrong). We could maybe clean that up as well.

yogeshbhutkar committed:
Remove redundant height property from background image control dropdown styles

yogeshbhutkar committed:
Refactor BackgroundImageControls and MediaReplaceFlow to improve button rendering and support custom toggle rendering

yogeshbhutkar commented:
As per the suggestions provided by @mirka, the following are the changes made in the recent commit:

1. Removed the code for `buttonVariant` abstract approach in favor of introducing the exposure of renderToggle as an optional prop.
2. Cleaned up `editMediaButtonRef` as it was referred not relevant in the recent implementations [here](https://github.com/WordPress/gutenberg/pull/68084#pullrequestreview-2515546731).
3. Removed name prop from being passed inside `BackgroundImageControls` component when using the `MediaReplaceFlow` as when renderToggle is provided, the name is also provided within the `Button`. This name is anyways supposed to be used by the `Button` itself, so we can save up some redundancy by passing it within renderToggle itself.

Here's a screencast post recent changes:
![Screen Recording Dec 20 2024](https://github.com/user-attachments/assets/27623f2a-26ba-4c95-b979-3aa6d0fa9c12)

The changes seem to work as expected on my end.

yogeshbhutkar committed:
Remove unnecessary blank line in MediaReplaceFlow component

tyxla commented on packages/block-editor/src/components/background-image-control/index.js:
```diff
@@ -369,15 +371,28 @@ function BackgroundImageControls( {
+						aria-haspopup="true"
+						onClick={ onToggle }
+						onKeyDown={ ( event ) => {
+							if ( event.keyCode === DOWN ) {
```
I know we've been doing it like that in other places, but `event.keyCode` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode). You might want to use:

```suggestion
							if ( event.key === 'ArrowDown' ) {
```

yogeshbhutkar replied:
Hi @tyxla, thanks for the review. I've incorporated the suggested changes in the latest commit.

yogeshbhutkar committed:
Update BackgroundImageControls to use 'ArrowDown' key for dropdown navigation

yogeshbhutkar committed:
Merge branch 'trunk' into 68028/fix-spacing-issue

Mamaduka commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -238,7 +242,6 @@ const MediaReplaceFlow = ( {
 								showSuggestions={ false }
 								onChange={ ( { url } ) => {
 									onSelectURL( url );
-									editMediaButtonRef.current.focus();
```
What's the reason for removing this focus handler?

yogeshbhutkar replied:
@Mamaduka, it's related to [one of the comments](https://github.com/WordPress/gutenberg/pull/68084#pullrequestreview-2515546731) mentioned above.

> Also as a separate issue, I noticed that the focus manipulation that editMediaButtonRef is being used to do is not relevant/appropriate anymore in these recent implementations.

Mamaduka commented on packages/block-editor/src/components/media-replace-flow/index.js:
```diff
@@ -137,21 +136,26 @@ const MediaReplaceFlow = ( {
 
+	const defaultRenderToggle = ( { isOpen, onToggle } ) => (
+		<ToolbarButton
+			aria-expanded={ isOpen }
+			aria-haspopup="true"
+			onClick={ onToggle }
+			onKeyDown={ openOnArrowDown }
```
I wonder if we should pass down these as default props for the media flow render toggle. This way, consumers don't have to re-implement a11y best practices.

mirka replied:
Exactly, that is what I had in mind with this [usage example](https://github.com/WordPress/gutenberg/pull/68084#pullrequestreview-2515546731). We can assume (and document) that the `renderToggle` function should accept and pass through `button` props to an eventual `button` element.

@yogeshbhutkar Let me know if you're unsure what to do here, I'd be happy to push up some proposed code.

Mamaduka replied:
@mirka, yeah, reread your example after I commented and noticed it 😅

mirka commented on packages/block-editor/src/components/background-image-control/index.js:
```diff
@@ -369,15 +370,28 @@ function BackgroundImageControls( {
 							displayInPanel,
 					} ),
 				} }
-				name={
-					<InspectorImagePreviewItem
-						className="block-editor-global-styles-background-panel__image-preview"
-						imgUrl={ url }
-						filename={ title }
-						label={ imgLabel }
-					/>
```
Because [`name`](https://github.com/WordPress/gutenberg/blob/f37e830d69bf225adb3e64f2127a4afff09d2d34/packages/block-editor/src/components/media-replace-flow/index.js#L54) is a standard prop of `MediaReplaceFlow`, I think it would be safer to continue using this prop rather than rendering it directly into our `renderToggle`.

One risk I'm thinking about is that the `MediaReplaceFlow` implementation could change some day to use the `name` prop in another place as well.

yogeshbhutkar committed:
Media Replace Flow: Add custom toggle support and fix button height

yogeshbhutkar committed:
added `renderToggle` prop details to readme

yogeshbhutkar committed:
Merge branch 'trunk' into 68028/fix-spacing-issue

yogeshbhutkar committed:
refactor: remove unused styles

yogeshbhutkar commented:
While implementing the suggested changes, I found that entirely removing [this CSS rule](https://github.com/WordPress/gutenberg/pull/68084#discussion_r1892711855) changes the height of the Image Preview component. Instead of completely removing it, the following rule will fix the issue:

https://github.com/WordPress/gutenberg/blob/e3df6a01c06d43cc5d96b3a72d6c32290892075b/packages/block-editor/src/components/background-image-control/style.scss#L27

| Before | After |
| ------- | ----- |
|![Screenshot 2024-12-27 at 10 50 17 AM](https://github.com/user-attachments/assets/7780036b-2d94-46bf-9fdc-0935fc147634)|![Screenshot 2024-12-27 at 10 37 59 AM](https://github.com/user-attachments/assets/49046088-e21f-4756-9e57-adeb02fb0f14)|


<hr />

Also, it looks like there were some missed lints in the JSON file that previously got committed to the latest trunk https://github.com/WordPress/gutenberg/blob/a4aac838a1b89330f0be2e785f5a9cd5aef8523a/packages/block-library/src/post-template/block.json#L53

The `pre-commit hook` seems to have fixed those instances (https://github.com/WordPress/gutenberg/pull/64425) in this PR while pushing the merged code to resolve conflicts.

<hr />

Finally, [this comment](https://github.com/WordPress/gutenberg/pull/68084#discussion_r1897892576) clarified the suggested approach further. So, I've updated the renderToggle implementation to accept button props, which must be passed to the button component. I've also documented this change.

If you feel this is not the best approach, I can refer to a small code example.

<hr />

Here's a screencast after the suggested changes:

![Recent Changes](https://github.com/user-attachments/assets/0ec4be4f-5fb0-460d-87ab-f487bc4dd8d6)

Thanks a lot to the reviewers for these constant follow-ups ✨.

CC: @Mamaduka @mirka

Mamaduka commented on packages/block-library/src/post-template/block.json:
These changes seem unrelated. Are they coming from the ESLint fix command?

Mamaduka commented on packages/block-editor/src/components/background-image-control/style.scss:
Are changes in this file still necessary?

yogeshbhutkar replied:
Yes, @Mamaduka. I've tried to explain it above in [this](https://github.com/WordPress/gutenberg/pull/68084#issuecomment-2563338352) comment.

Thanks a lot for reviewing the PR 🚀

mirka commented on packages/block-editor/src/components/background-image-control/style.scss:
```diff
@@ -23,7 +23,10 @@
-		height: 36px;
+
+		.block-editor-global-styles-background-panel__dropdown-toggle {
+			height: 36px;
```
```suggestion
			height: 40px;
```

yogeshbhutkar replied:
Thanks for the review @mirka. I've updated the height in the latest commit.

![Screenshot 2024-12-30 at 11 48 30 AM](https://github.com/user-attachments/assets/8328390b-299a-42e9-a292-445f5c1886c4)

mirka reviewed (commented):
API changes look great, thank you! I think we're good to go once the height value is fixed.

yogeshbhutkar committed:
style: increase dropdown toggle height to 40px

mirka reviewed (approved):
Thanks for all the follow-ups! 🚀
